File: backend/src/main/java/com/tecnoo/helpdesk/Models/Ticket.java
package com.tecnoo.helpdesk.Models;

import lombok.Data;

import javax.persistence.CollectionTable;
import javax.persistence.Column;
import javax.persistence.ElementCollection;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.PrePersist;
import javax.persistence.Table;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;

import org.hibernate.annotations.GenericGenerator;

import com.tecnoo.helpdesk.Models.Enums.Prioridade;
import com.tecnoo.helpdesk.Models.Enums.Status;
import com.tecnoo.helpdesk.Models.Enums.TipoAtendimento;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Data
@Entity
@Table(name = "tbchamados")
public class Ticket {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id_chamado")
    private Long id;

    @NotBlank(message = "O campo protocolo é obrigatório")
    @GeneratedValue(generator = "uuid")
    @GenericGenerator(name = "uuid", strategy = "uuid2")
    @Column(unique = true)
    private String protocolo;

    private String titulo;

    @NotNull
    @ManyToOne
    private Cliente cliente;

    private LocalDateTime dataAbertura = LocalDateTime.now();

    private LocalDateTime dataFinalizacao;

    private LocalDateTime dataLimite;

    private Status status = Status.ABERTO;

    private Prioridade prioridade;

    private TipoAtendimento tipoAtendimento;

    private String descricao;

    @ElementCollection
    @CollectionTable(name = "tbchamados_imagens", joinColumns = @JoinColumn(name = "id_chamado"))
    @Column(name = "image_url")
    private List<String> imageUrls = new ArrayList<>();

    @OneToMany(mappedBy = "chamado")
    private List<HistoricoChamado> historicos = new ArrayList<>();

    @PrePersist
    public void prePersist() {
        this.protocolo = generateProtocol();
    }

    private String generateProtocol() {
        LocalDateTime now = LocalDateTime.now();

        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("ddMMyyyyHHmmss");

        String formatedDate = now.format(formatter);

        String uuid = UUID.randomUUID().toString();

        return "TI-" + formatedDate + "-" + uuid.substring(0, 4);
    }
}

File: front_mobile/lib/services/models/ticket/ticket.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first
import 'package:brasil_fields/brasil_fields.dart';
import 'package:front_mobile/services/models/historico_chamado/historico.dart';

class Ticket {
  final int id;
  final String? protocolo;
  final String? titulo;
  final String? descricao;
  final int idCliente;
  final DateTime dataAbertura;
  final DateTime? dataFinalizacao;
  final DateTime? dataLimite;
  final String? status;
  final String? prioridade;
  final String? tipoAtendimento;
  final List<HistoricoChamadoModel?>? historicos;

  Ticket({
    required this.id,
    this.protocolo,
    this.titulo,
    this.descricao,
    required this.idCliente,
    required this.dataAbertura,
    this.dataFinalizacao,
    this.dataLimite,
    this.status,
    this.prioridade,
    this.tipoAtendimento,
    this.historicos,
  });

  factory Ticket.fromJSON(Map<String, dynamic> json) {
    List<HistoricoChamadoModel?>? historicos = [];

    if (json['historicos'] != null) {
      json['historicos'].forEach((v) {
        historicos.add(HistoricoChamadoModel.fromJson(v));
      });
    }

    return Ticket(
        id: json['id'],
        protocolo: json['protocolo'],
        titulo: json['titulo'],
        descricao: json['descricao'],
        idCliente: json['idCliente'],
        dataAbertura: parseDateTime(json['dataAbertura'])!,
        dataFinalizacao: parseDateTime(json['dataFinalizacao']),
        dataLimite: parseDateTime(json['dataLimite']),
        status: json['status'],
        prioridade: json['prioridade'],
        tipoAtendimento: json['tipoAtendimento'],
        historicos: historicos);
  }

  @override
  String toString() {
    return 'Ticket(id: $id, protocolo: $protocolo, titulo: $titulo, descricao: $descricao, idCliente: $idCliente, dataAbertura: $dataAbertura, dataFinalizacao: $dataFinalizacao, dataLimite: $dataLimite, status: $status, prioridade: $prioridade, tipoAtendimento: $tipoAtendimento, historicos: $historicos)';
  }
}

DateTime? parseDateTime(dynamic value) {
  return value != null ? UtilData.obterDateTimeHora(value) : null;
}
